src/pandas/array.h

```diff
@@ -20,9 +20,19 @@ class Array {
  public:
   virtual ~Array() {}
 
-  int64_t length() const { return length_; }
-  std::shared_ptr<DataType> type() const { return type_; }
-  DataType::TypeId type_id() const { return type_->type(); }
+  virtual int64_t length() const = 0;
+  // There are two methods to obtain the data type.
+  // The signature without a shared_ptr allows sub-classes
+  // to have a covariant return type, which eliminates the
+  // need/danger of doing a static_cast when dealing with
+  // a concrete sub-class. Ideally, the shared_ptr signature
+  // would suffice, but the compiler cannot treat a shared_ptr
+  // to a base class and a shared_ptr to a subclass as a
+  // covariant return type.
+  virtual TypePtr type() const = 0;
+  virtual const DataType& type_reference() const = 0;
```

Comment:
If you go the TypeId route, then you don't really need the visitor pattern. However, if you go with the visitor pattern, when the visitor visits a concrete Array (e.g. - CategoryArray), the type_reference function makes it possible to get the concrete DataType (e.g. - CategoryType). That is easier that doing yet another double dispatch on the DataType and safer than the user of the API doing a static_cast to the type they think corresponds to the Array type. In general, the visitor pattern and the covariant return type will give you the most compile time safety.

In regards to the visitor pattern vs. using TypeId's with switch statements, I think it's best to test the performance. If there is a really compelling performance reason, then it may be worthwhile to use TypeId's in favor of the visitor pattern.